specs/fizzbee/LeaderFollower/VeniceLeaderFollower.fizz

```diff
@@ -0,0 +1,136 @@
+#   not modelled in this spec.
+# - Clients publish messages to the real-time topic, which is consumed by the leader.
+# - The leader appends the consumed messages to the version topic.
+# - The followers consume messages from the version topic.
+#
+# The protocol is correct if the persisted records of all nodes are eventually consistent.
+
+# Note: The comments in this spec are written as a tutorial for the FizzBee language.
+# In a real example, the comments should describe why you do what you do, not what you do.
+
+# Assertions are specified with assertion keyword. The body of the assertion is similar
+# to a python function returning a boolean.
+always eventually assertion ReplicasConsistent:
+    for i in NODE_IDS:
+        for j in range(i + 1, len(nodes)):
+            # Check if the persisted records of the current pair of nodes are equal
+            if nodes[i]['persistedRecords'] != nodes[j]['persistedRecords']:
+                return False
+
+    return True
+
+# Init is a special action called when the model starts. It is used to initialize the state of the model.
+action Init:
+    # Currently, FizzBee does not support parameters. So, we have to define these as variables with uppercase.
+    KEYS = ('key1', 'key2')
+    VALUES = ('value1', 'value2')
+    LEADER = 'LEADER'
+    FOLLOWER = 'FOLLOWER'
+    N_NODES = 3
+    NODE_IDS = range(N_NODES)
+
+    # Initialize the state of each node
+    real_time_topic = []
+    version_topic = []
+    nodes = [
+        {
+            'state': 'FOLLOWER',
+            'rtOffset': 0,
+            'vtOffset': 0,
+            'persistedRecords': {}
+        }
+        for i in NODE_IDS
+    ]
+
+
+# The action keyword is used to define an action. The atomic keyword implies
+# there are no yield points within the statements.
+atomic action PromoteLeader:
+    # A FOLLOWER ca be promoted to be a LEADER if and only if there is no leader in the system.
+
+    if [node for node in nodes if node['state'] == 'LEADER']:
+        return
+    # `any` is a special keyword that creates non-determinism or alternate timelines.
+    # This is similar in structure to `for`, but in each timeline, only one of the
+    # non-deterministic choices is executed.
+    any id in NODE_IDS:
+        nodes[id]['state'] = 'LEADER'
+
+
+atomic action DemoteLeader:
+    any id in NODE_IDS:
+        if nodes[id]['state'] == 'LEADER':
+            nodes[id]['state'] = 'FOLLOWER'
+
+
+atomic action ClientProducesToVenice:
+    any kv in [(k, v) for k in KEYS for v in VALUES]:
+        real_time_topic.append(kv)
+
+
+# The fair keyword implies this action is weakly fair. That is, the action is always enabled,
+# it will eventually happen.
+atomic fair action LeaderConsume:
+  any id in NODE_IDS:
+    if nodes[id]['state'] == 'LEADER':
+        if nodes[id]['vtOffset'] >= len(version_topic):
+            real_time_consume(id)
+        else:
+            version_topic_consume(id)
```

Comment:
This seems to mean: "if the leader's VT offset is below the end offset of the VT, then consume the VT, else consume the RT"

This makes sense I suppose, though in practice (within the Venice code) this particular decision only happens as part of the transition from follower to leader, and once it is determined ("somehow") that the leader can begin consuming from the RT, it will not go back on this decision (unless it is demoted to follower). In this spec, however, I don't see this being a "sticky" decision, it seems to be evaluated every time.

I guess the question is not necessarily for @jayaprabhakar but perhaps for @ZacAttack, since Jay may be just translating what is spelled out in the TLA+ spec. Zac, does the above make sense? Is the spec intentionally modeling a simplification of the real logic, or does this somehow map precisely to the system's behavior, and I'm just missing some detail?

Reply:
The above makes sense.  The TLA+ spec as written is currently evaluating weather or not it's caught up based on this, and it was written that way at the time as a simplification of the Venice code.

It might be more interesting to instead model it as a state.  Like, the state of a node is either caught up or not caught up, and that state flips only after it's read through the VT to topic switch.

Reply:
Thanks. I noticed this. In the current spec, if `[nodeId].rtOffset`  (`nodes[nodeId]['rtOffset']`) of the leader is updated but not the `[nodeId].vtOffset` (`nodes[nodeId]['vtOffset']`). That would imply, when a leader is demoted, the newly demoted node would take a long time to catchup. So, the two ways to avoid slow and unnecessary catchup, was either to doing it continuously like the current spec, or to use a local state variable. I wasn't sure what the design decision was.

So I decided to translate the TLA+ spec as it is. 

From the discussion, I assume, we just need to have a state variable to mark the follower as caught up. 
(FYI, changing it by updating vtoffset + local state reduces the possible states and the model checking time significantly)

Reply:
In that case, when the newly promoted leader catches up with the version topic, it still starts to consume real time topic from where that node left of last when it got demoted, instead of picking the offset from the latest version topic message.

That is clearly impractical. Can I also change that to use the latest version topic?

Reply:
Yeah, a state variable is good enough.  The way it works in production is the follower just keeps a highwatermark based on what's it's consumed from VT.  So the offset from the latest version topic I think is good.